### ISSUE TITLE
Removing static from ssl_parse_certificate_chain

### ISSUE THREAD

**Issue by Dragris** [Mbed-TLS/mbedtls#6892]
### Suggested enhancement
Remove the static from the ssl_parse_certificate_chain method found in ssl_tls.c.
Alternatively, a method that gives the peer's certificate from the last tried connection could be added.

### Justification

Mbed TLS needs this because if the peer certificate from a failed handshake wants to be retrieved and said handshake fails during the MBEDTLS_SSL_SERVER_CERTIFICATE step, such as in the case of a Man In The Middle attack, the certificate cannot be retrieved.

**Comment by tom-daubney-arm** [comment]
Hi @Dragris ,
Thanks for raising an issue. I would like to get some more details from you please:

1) Could you please clarify the high-level problem you are running into? And following on from that, why do you think removing ssl_parse_certificate_chain will help solve that problem?

2) Are you aware of the following functions: mbedtls_ssl_get_peer_cert(), mbedtls_ssl_conf_verify(), mbedtls_ssl_conf_authmode() and could you use any of these to help with the problem you are facing?

**Comment by Dragris** [comment]
Hi @tom-daubney-arm,

what I'm trying to do is to retrieve the peer certificate when the connection is under a MITM attack. To be precise, this MITM attack is being detected during the handshake. I've tried using the mbedtls_ssl_get_peer_cert() in this circumstances but I got an empty certificate.

Removing the static scope from the function could be helpful in this case, as after the failed handshake the required ssl_context is still present and the peer certificate can be obtained through the method.

Regarding the other two methods that you suggest, I will take a look at them to see if they're useful for in this scenario.

Thank you a lot for the answer!

**Comment by tom-daubney-arm** [comment]
Hi @Dragris ,

Thanks for the explanation. Please have a look at the functions mentioned and see if they can help you. Please come back and let me know your findings and then if there is still an issue present we can make sure it is logged correctly.

**Comment by paul-elliott-arm** [comment]
Hi!
`mbedtls_ssl_conf_verify()` can setup a a certificate verification callback, which should give you access to the cert even if the handshake fails after this. I'm not sure what you mean about changing configuration to be simply ignored or having to do a second handshake, the cert was passed as part of the first handshake, why would you need to do another?

Clearly we need to change at least the docs on `mbedtls_ssl_get_peer_cert()`, and I will open another ticket for that.

**Comment by tom-daubney-arm** [comment]
Hi @Dragris ,

Any update on this? I will be closing the issue in 24 hours if I do not hear back. Thanks in advance.

**Comment by Dragris** [comment]
Hello, 
I'm retesting the `mbedtls_ssl_conf_verify()` method and I got it wrong the first time. Indeed it can be useful to get a pointer to the certificate before the handshake fails. However I've noticed that after the handshake fails the data being pointed is all lost. I don't know if there is a function that would let me copy the contents of the certificate or if it is advised to perform my operation with said certificate within the callback.

Apologies for the misunderstanding from the first suggestion.
